app/services/patient_import/spreadsheet_transformer.rb

```diff
@@ -0,0 +1,304 @@
+      drug_names = [
+        row[:first_visit_medication_1],
+        row[:first_visit_medication_2],
+        row[:first_visit_medication_3],
+        row[:first_visit_medication_4],
+        row[:first_visit_medication_5]
+      ]
+
+      drugs = drug_names.map { |name|
+        medication(
+          name: name,
+          patient_id: patient_id,
+          created_at: row[:first_visit_date]
+        )
+      }.compact
+
+      # Delete the first drugs if last drugs are available
+      if last_prescription_drugs(row, patient_id: patient_id).any?
+        drugs.each do |drug|
+          drug.merge!(
+            is_deleted: true,
+            updated_at: timestamp(row[:last_visit_date]),
+            deleted_at: timestamp(row[:last_visit_date])
+          )
+        end
+      end
+
+      drugs
+    end
+
+    def last_prescription_drugs(row, patient_id:)
+      drug_names = [
+        row[:last_visit_medication_1],
+        row[:last_visit_medication_2],
+        row[:last_visit_medication_3],
+        row[:last_visit_medication_4],
+        row[:last_visit_medication_5]
+      ]
+
+      drug_names.map { |name|
+        medication(
+          name: name,
+          patient_id: patient_id,
+          created_at: row[:last_visit_date]
+        )
+      }.compact
+    end
+
+    def patient_id(row)
+      PatientBusinessIdentifier.find_by(
+        identifier_type: row[:identifier_type],
+        identifier: row[:identifier]
+      )&.patient_id || SecureRandom.uuid
+    end
+
+    def business_identifier_metadata_version(row)
+      case row[:identifier_type]
+      when "simple_bp_passport"
+        "org.simple.bppassport.meta.v1"
+      else
+        "org.simple.#{row[:identifier_type]}.meta.v1"
+      end
+    end
+
+    def registration_facility_id
+      facility.id
+    end
+
+    def import_user
+      @import_user = PatientImport::ImportUser.find_or_create
+    end
+
+    def create_import_user
+      user = User.new(
+        full_name: "import-user",
+        organization_id: Organization.take,
+        device_created_at: Time.current,
+        device_updated_at: Time.current
+      )
+      phone_number_authentication = PhoneNumberAuthentication.new(
+        phone_number: IMPORT_USER_PHONE_NUMBER,
+        password: "#{rand(10)}#{rand(10)}#{rand(10)}#{rand(10)}",
+        registration_facility_id: facility.id
+      ).tap do |pna|
+        pna.set_otp
+        pna.invalidate_otp
+        pna.set_access_token
+      end
+
+      user.phone_number_authentications = [phone_number_authentication]
+      user.sync_approval_denied("bot user for import")
+      user.save!
+
+      user
+    end
+
+    def patient_status(row)
+      row[:died] == "yes" ? :dead : :active
+    end
+
+
+
+    def medication(name:, patient_id:, created_at:)
+      medication_record = medication_by_name_dosage_and_frequency(name) || medication_by_name_and_dosage(name)
+
+      return nil unless medication_record.present?
```

Comment:
IT's not as bad as it sounds, because the XLS files we hand out have validations in these data fields. It's also incredibly easy to use if it's a single field, instead of separate fields for each ☝️ 

This line actually catches cells that still have `None` in them 😅 and also, yes, ignores any invalid data.

Reply:
Cool, good to know.